core/remotes/docker/pusher.go

```diff
@@ -178,17 +178,21 @@ func (p dockerPusher) push(ctx context.Context, desc ocispec.Descriptor, ref str
 			// query and send the request again.
 			resp, err = preq.doWithRetries(pctx, nil)
 			if err != nil {
-				return nil, err
+				if !errors.Is(err, ErrInvalidAuthorization) {
```

Comment:
Does this appear on push failure, or when we couldn't get the token scopes requested from `ContextWithAppendPullRepositoryScope`. Ideally we could error directly when we can't get the token for the scope that is in `ctx` and fail with a typed error, so we can capture it in here more precisely.

Reply:
As `selectRepositoryMountCandidate` picks only one candidate source (looks to be random), a low-priority follow-up could be to try again with another candidate if it exists.

Reply:
This seems right, my understanding of the code is if the registry sets the `error` parameter, then the typed error will be returned and the 401 response case below will not be hit.

I am curious though, which registry is returning a 401 in this case rather than a 202. I wonder if 403 should be considered here, it doesn't seem to be called out in the OCI spec and not sure any registry returns that though.

Reply:
> I am curious though, which registry is returning a 401 in this case rather than a 202.

I'm pretty sure Docker Hub is one of such registries. Even though when I look at the spec then there is indeed a `SHOULD` clause to ignore it if it can't continue.

Reply:
@tonistiigi The change in this PR allows the `Authorizer` to return an error from `Authorize` or `AddResponses`. Returning an error lets the `Authorizer` stop any further retries. I've updated the tests to cover both scenarios. 

`selectRepositoryMountCandidate` seems to pick the candidate with the longest common prefix (not a random one). I agree that we could also fall back to trying the whole list, but I'm not sure if that's in scope for this PR or if it would be better to do it as a follow-up.